tao_compiler/mlir/disc/transforms/disc_lhlo_rewriter.cc

```diff
@@ -0,0 +1,156 @@
+    bool is_shape_equal = true;
+    // check all input operand, the last one is output buffer
+    for (int i = 0; i < operands - 2; ++i) {
+      is_shape_equal &=
+          shape_analysis.isShapeEqual(op->getOperand(i), op->getOperand(i + 1));
+    }
+    return is_shape_equal;
+  }
+
+  LogicalResult matchAndRewrite(lmhlo::ConcatenateOp lhloOp,
+                                PatternRewriter& rewriter) const override {
+    Operation* op = lhloOp.getOperation();
+    if (!isFixedShape(lhloOp)) return failure();
+
+    auto operands = op->getOperands();
+    int num_input_operands = op->getNumOperands() - 1;
+
+    SmallVector<Value, 4> ptr_array;
+    auto ptr_type = IntegerType::get(op->getContext(), 64);
+    SmallVector<Value, 4> ins(operands.begin(), operands.end());
+    auto out = ins.back();
+    auto ptr_alloc =
+        rewriter
+            .create<memref::AllocOp>(
+                op->getLoc(), MemRefType::get({operands.size()}, ptr_type))
+            .getResult();
+    for (int i = 0; i < num_input_operands; ++i) {
+      Value idx = rewriter.create<arith::ConstantIndexOp>(op->getLoc(), i);
+      auto ptr = rewriter.create<disc_ral::GetPointerOp>(op->getLoc(), ptr_type,
+                                                         op->getOperand(i));
+      ptr->setAttr(kDiscPlaceAssignment, rewriter.getStringAttr(kGpu));
+      rewriter.create<memref::StoreOp>(op->getLoc(), ptr, ptr_alloc, idx);
+    }
+    auto device_ptr_alloc =
+        rewriter
+            .create<memref::AllocOp>(
+                op->getLoc(), MemRefType::get({operands.size()}, ptr_type))
+            .getResult();
+    rewriter.create<lmhlo_disc::H2DOp>(op->getLoc(), ptr_alloc,
+                                       device_ptr_alloc);
+    // {inputs, input_ptr, out}
+    ins.insert(ins.begin() + num_input_operands, device_ptr_alloc);
+    rewriter.create<lmhlo_disc::ConcatenateOp>(op->getLoc(), llvm::None, ins,
+                                               op->getAttrs());
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct DiscLhloRewriterPass
+    : public DiscLhloRewriterPassBase<DiscLhloRewriterPass> {
+  using DiscLhloRewriterPassBase<
+      DiscLhloRewriterPass>::DiscLhloRewriterPassBase;
+
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<lmhlo_disc::LmhloDiscDialect, memref::MemRefDialect,
+                    disc_ral::RalDialect, lmhlo::LmhloDialect>();
+  }
+
+ public:
+  DiscLhloRewriterPass() = default;
+
+  void runOnOperation() override {
+    auto& context = getContext();
+    RewritePatternSet patterns(&context);
+    ConversionTarget target(context);
+    target.addLegalDialect<arith::ArithDialect, lmhlo_disc::LmhloDiscDialect,
+                           memref::MemRefDialect, shape::ShapeDialect,
+                           tensor::TensorDialect>();
+    target.addIllegalOp<lmhlo::ConcatenateOp>();
+
+    patterns.insert<HloToLhloConcatenateOpConverter>(&context);
```

Comment:
Are there some bad cases of this conversion? As there is no way to turn it off, for example with a flag.

Reply:
it only works on the fixed shape, maybe we can open it by default if this PR passed all the unit test? Because there are too much  flags in the current codebase 😅